packages/flutter/lib/src/rendering/object.dart

```diff
@@ -378,28 +389,51 @@ class PaintingContext extends ClipContext {
 
   /// Clip further painting using a rectangle.
   ///
+  /// {@template flutter.rendering.object.needsCompositing}
   /// * `needsCompositing` is whether the child needs compositing. Typically
-  ///   matches the value of [RenderObject.needsCompositing] for the caller.
+  ///   matches the value of [RenderObject.needsCompositing] for the caller. If
+  ///   false, this method returns null, indicating that a layer is no longer
+  ///   necessary. If a render object calling this method stores the `oldLayer`
+  ///   in its [RenderObject.layer] field, it should set that field to null.
+  /// {@end template}
   /// * `offset` is the offset from the origin of the canvas' coordinate system
   ///   to the origin of the caller's coordinate system.
   /// * `clipRect` is rectangle (in the caller's coordinate system) to use to
   ///   clip the painting done by [painter].
   /// * `painter` is a callback that will paint with the [clipRect] applied. This
   ///   function calls the [painter] synchronously.
   /// * `clipBehavior` controls how the rectangle is clipped.
-  void pushClipRect(bool needsCompositing, Offset offset, Rect clipRect, PaintingContextCallback painter, { Clip clipBehavior = Clip.hardEdge }) {
+  /// {@template flutter.rendering.object.oldLayer}
+  /// * `oldLayer` is the layer created in the previous frame. Specifying the
+  ///   old layer gives the engine more information for performance
+  ///   optimizations. Typically this is the value of [RenderObject.layer] that
+  ///   a render object creates once, then reuses for all subsequent frames
+  ///   until a layer is no longer needed (e.g. the render object no longer
+  ///   needs compositing) or until the render object changes the type of the
+  ///   layer (e.g. from opacity layer to a clip rect layer).
+  /// {@end template}
+  ClipRectLayer pushClipRect(bool needsCompositing, Offset offset, Rect clipRect, PaintingContextCallback painter, { Clip clipBehavior = Clip.hardEdge, ClipRectLayer oldLayer }) {
     final Rect offsetClipRect = clipRect.shift(offset);
     if (needsCompositing) {
-      pushLayer(ClipRectLayer(clipRect: offsetClipRect, clipBehavior: clipBehavior), painter, offset, childPaintBounds: offsetClipRect);
+      ClipRectLayer layer = oldLayer;
+      if (layer == null) {
+        layer = ClipRectLayer(clipRect: offsetClipRect, clipBehavior: clipBehavior);
+      } else {
+        layer
```

Comment:
Shall we also invalidate the layer's engineLayer at this point?

Reply:
I think that's layer-specific. The old `engineLayer` may still be needed to be passed as `oldLayer` during `compositeFrame`.

Reply:
I see. So `engineLayer` could either be:
1. in a state that's synchronized with the layer so we can directly use it in `addRetained`, 
2. in a stale state where one can only use it as `oldLayer`.

Is there some mechanism that we can enforce the developer to not misuse such two states of the `engineLayer`?

Reply:
Yes, I added a bunch of assertions in the `SceneBuilder` in the engine code.

Reply:
LeaderLayer shouldn't even have an engine layer, should it?

Reply:
@yjbanov ^

Reply:
`LeaderLayer` sometimes creates a `TransformEngineLayer`:

https://github.com/flutter/flutter/blob/b11c5acaa6200a8d662705a312a2e5fb7714fe2e/packages/flutter/lib/src/rendering/layer.dart#L1755

Reply:
Oh, right.